### PR TITLE
Add support for ltex-ls client

### PR DESCRIPTION
**Pull request by Javier-varez**
ltex-ls uses the `data` field in requests and
needs to have complete support for it. It turns
out that ltex-ls also requires a fix on the
client side because they publish the HTTP request
with the wrong Content-Type, so axum does not
even forwards the request to check. This is fixed
by:

https://github.com/Javier-varez/ltex-ls/commit/bb116c7dbac199b10d5df25bdadd878aa8dc72ca

### CODE REVIEW

File: src/api.rs [hunk 1]
```diff
@@ -48,8 +48,114 @@ mod test {
...
-    pub text: String,
+#[serde(untagged)]
+#[serde(rename_all = "camelCase")]
+pub enum AnnotationElement {
+    Text {
+        text: String,
+    },
+    Markup {
+        markup: String,
+        interpret_as: Option<String>,
```
Comment (cpg314) [inline src/api.rs:63]:
```suggestion
        // Interpret the markup as this string for analysis.
        // E.g. "\n\n" when `markup` is `<p>`
        interpret_as: Option<String>,
```

File: src/api.rs [hunk 2]
```diff
@@ -48,8 +48,114 @@ mod test {
...
+                AnnotationElement::Markup { markup, .. }
+                    if !markup.is_empty() && markup.trim().is_empty() =>
+                {
+                    markup.len()
+                }
+                AnnotationElement::Markup { interpret_as, .. } => {
+                    interpret_as.as_deref().unwrap_or("").len()
+                }
+            })
+            .reduce(|acc, v| acc + v)
```
Comment (cpg314) [inline src/api.rs]: Wouldn't `sum()` work here?

File: src/api.rs [hunk 3]
```diff
@@ -48,8 +48,114 @@ mod test {
...
+            .unwrap_or(0)
+    }
+
+    /// Translate a textual span into the span of the text containing markup
+    pub fn translate_span(&self, start: usize, end: usize) -> (usize, usize) {
+        let mut text_offset = 0;
+        let mut markup_offset = 0;
+
+        let mut mapped_start = None;
```
Comment (cpg314) [inline src/api.rs]: It would be nice to annotate the types here for readability.

File: src/api.rs [hunk 4]
```diff
@@ -48,8 +48,114 @@ mod test {
...
+                {
+                    (markup.len(), markup.len())
+                }
+                AnnotationElement::Markup {
+                    markup,
+                    interpret_as,
+                } => (interpret_as.as_deref().unwrap_or("").len(), markup.len()),
+            };
+
+            if mapped_start.is_none()
```
Comment (cpg314) [inline src/api.rs]: I assume could also be written with a `for mapped in [(&mut mapped_start, start), (&mut mapped_end, end))]`

File: src/api.rs [hunk 5]
```diff
@@ -76,12 +182,15 @@ impl Request {
             name: "".into(),
         }
     }
-    pub fn text(self) -> Option<String> {
+
+    pub fn annotations(self) -> Option<Annotations> {
```
Comment (cpg314) [inline src/api.rs]: I think we should propagate the result here, as the Option only makes it difficult to debug invalid queries, like I just did.

**Review by cpg314 — COMMENTED** [review]
Thanks, that's great! I'm happy to see it can be useful to someone else :)

It seems that the change breaks with the browser extension, which sends the following query:
```
[2023-10-06 21:57:26] DEBUG [ltapiserv_rs] Query Request {
    text: None,
    data: Some(
        "{\"text\":\"Thanks, that's great!\\n\\nHowever, it seems that it breaks with the browser extension:\"}",
    ),
    language: "auto",
}
```
The problem happens when deserializing the data field: apparently it can be either `{"annotations": ...}` or `{"text": ...}`. 
That's the kind of [undocumented](https://languagetool.org/http-api/swagger-ui/#!/default/post_check) (I think) behaviours I had to discover to get everything to work.

File: src/api.rs (hunk 2 above)
Comment (Javier-varez) [reply src/api.rs]: Yes, it would! It would help if I was more familiar with some of these methods...

File: src/api.rs (hunk 4 above)
Comment (Javier-varez) [reply src/api.rs]: I went for a closure instead, but it achieves the same in practice.

File: src/api.rs (hunk 5 above)
Comment (Javier-varez) [reply src/api.rs]: I agree, I also needed it  when debugging the integration with ltex-ls

**Comment by Javier-varez** [comment]
Thank you for writing this tool in the first  place! It is really helpful, I must say! 

Indeed, I wondered why the data field had a text field inside, but didn't think of testing with the browser plugins. It should be working now.

**Comment by cpg314** [comment]
Thanks, merged. I'll add some tests with documentation about tested clients, and release a new version.

**Pull request closed without merging**